File: apis/python_interface_helpers/stk_options.py
class stk_options:
	def __init__(self,opts_str,_opts=None):
		if _opts != None:
			self._opts = _opts
		else:
			if type(opts_str) is dict:
				opts_str = self.hash_to_string(opts_str)
			self._opts = stk_build_options(opts_str,len(opts_str))
	def ref(self):
		return self._opts
	def find_option(self,name):
		opt = stk_find_option(self._opts,name,None);
		if opt != None:
			return stk_options(None,stk_void_to_options_t(opt))
		else:
			return None
	def find_sub_option(self,name):
		subopts = stk_find_option(self._opts,name,None);
		return stk_void_to_options_t(subopts);
	def free_built_options(self,opts):
		if opts != None:
			stk_free_built_options(opts)
		else:
			stk_free_built_options(self._opts)
	def free_sub_option(self,optname):
		self.free_built_options(self.find_sub_option(optname))
	def close(self):
		if self._opts != None:
			self.free_built_options(None)
			stk_free_options(self._opts)
			self._opts = None
	def append_dispatcher_wakeup_cb(self):
		self._opts = stk_append_dispatcher_wakeup_cb(self._opts)
	def remove_dispatcher_wakeup_cb(self):
		if self._opts != None:
			stk_clear_cb(self._opts,"wakeup_cb")
	def append_dispatcher_fd_cbs(self,data_flow_group):
		self._opts = stk_option_append_dispatcher_fd_cbs(data_flow_group,self._opts)
	def remove_dispatcher_fd_cbs(self):
		if self._opts != None:
			stk_clear_cb(self._opts,"fd_created_cb")
			stk_clear_cb(self._opts,"fd_destroyed_cb")
	def remove_data_flow(self,name):
		if self._opts != None:
			stk_clear_cb(self._opts,name)
	def append_dispatcher(self,dispatcher):
		self._opts = stk_append_dispatcher(self._opts,dispatcher)
	def append_data_flow(self,data_flow_group,df):
		self._opts = stk_append_data_flow(self._opts,data_flow_group,df.ref())
	def remove_data_flow(self,data_flow_group):
		if self._opts != None:
			stk_clear_cb(self._opts,data_flow_group)
	def append_sequence(self,option_name,sequence):
		self._opts = stk_append_sequence(self._opts,option_name,sequence.ref())
	def remove_sequence(self,option_name):
		if self._opts != None:
			stk_clear_cb(self._opts,option_name)
	def stk_add_callback_option(self,name,cb):
		return None
	def update_ref(self,newref):
		self._opts = newref
	@classmethod
	def hash_to_string(cls,d,indent=0):
		s = ""
		for k, v in d.iteritems():
			if type(v) is dict:
				for _ in range(indent):
					s = s + " "
				s = s + str(k) + "\n"
				s = s + stk_options.hash_to_string(v,indent+1)
			else:
				for _ in range(indent):
					s = s + " "
				s = s + str(k) + " " + str(v) + "\n"
		return s

File: apis/ruby_api/stk_options.rb
require 'stkoptions'

class Stk_options
	@_opts = nil
	def ref()
		@_opts
	end

	def initialize(opts_str,opts=nil)
		if !opts.nil?
			@_opts = opts
		else
			opts_str = Stk_options.hash_to_string(opts_str) if opts_str.is_a?(Hash)
			@_opts = Stkoptions.stk_build_options(opts_str,opts_str.length)
		end
	end

	def find_option(name)
		opt = Stkoptions.stk_find_option(@_opts,name,nil);
		if !opt.nil?
			return Stk_options.new(nil,Stkoptions.stk_void_to_options_t(opt))
		else
			return nil
		end
	end

	def find_sub_option(name)
		Stkoptions.find_sub_option(@_opts,name)
	end

	def free_built_options(opts)
		if opts
			Stkoptions.stk_free_built_options(opts)
		else
			Stkoptions.stk_free_built_options(@_opts)
		end
	end

	def close()
		Stkoptions.stk_free_options(@_opts) if @_opts
	end

	def append_sequence(option_name,sequence)
		@_opts = Stkoptions.stk_append_sequence(@_opts,option_name,sequence.ref())
	end

	def update_ref(newref)
		@_opts = newref
	end

	def append_dispatcher(dispatcher)
		@_opts = Stkoptions.stk_append_dispatcher(@_opts,dispatcher)
	end

	def append_data_flow(data_flow_group,df)
		@_opts = Stkoptions.stk_append_data_flow(@_opts,data_flow_group,df.ref())
	end

	def remove_data_flow(name)
		Stkoptions.stk_clear_cb(@_opts,name) if @_opts
	end

	def append_dispatcher_fd_cbs(data_flow_group)
		@_opts = Stkoptions.stk_option_append_dispatcher_fd_cbs(data_flow_group,@_opts)
	end

	def remove_dispatcher_fd_cbs()
		if @_opts
			Stkoptions.stk_clear_cb(@_opts,"fd_created_cb")
			Stkoptions.stk_clear_cb(@_opts,"fd_destroyed_cb")
		end
	end

	def self.hash_to_string(h,indent=0)
		s = ""
		h.each do |k,v|
			if v.is_a?(Hash)
				indent.times { s = s + " " }
				s = s + "#{k.to_s}\n"
				s = s + self.hash_to_string(v,indent+1)
			else
				indent.times { s = s + " " }
				s = s + "#{k.to_s} #{v}\n"
			end
		end
		s
	end
end

